firmware/main.cpp

```diff
@@ -0,0 +1,377 @@
+}
+
+// Update LEDS in non-ISR context
+void update_leds() {
+    leds.updateLEDs(&state);
+}
+
+void send_can_message(bool drs, bool upshift, bool downshift, bool launch_ctl);
+
+void send_can() {
+    ScopedLock <Mutex> lock(state.mutex);
+    send_can_message(state.drsEngaged, false, false, state.launchCtlEngaged);
+}
+
+// ISR handler for LED loop
+void led_loop_handler(){
+    queue.call(update_leds);
+}
+
+void can_send_handler(){
+    queue.call(send_can);
+}
+
+// ISR callback functions for button test mode
+void upshift_test_cb_r(){ state.buttons[0] = false;  }
+void upshift_test_cb_f(){ state.buttons[0] = true; }
+void downshift_test_cb_r(){ state.buttons[1] = false;  }
+void downshift_test_cb_f(){ state.buttons[1] = true; }
+void drs_test_cb_r(){ state.buttons[2] = false;  }
+void drs_test_cb_f(){ state.buttons[2] = true; }
+void settings_test_cb_r(){ state.buttons[3] = false;  }
+void settings_test_cb_f(){ state.buttons[3] = true; }
+void aux1_test_cb_r(){ state.buttons[4] = false;  }
+void aux1_test_cb_f(){ state.buttons[4] = true; }
+void aux2_test_cb_r(){ state.buttons[5] = false;  }
+void aux2_test_cb_f(){ state.buttons[5] = true; }
+
+// Enter into button test mode and configure IRS callbacks for buttons
+void test_mode() {
+    printf("Entering button test mode.\n");
+    isTestMode = true;
+    leds.updateState(ButtonTest);
+    upshift_btn.rise(callback(upshift_test_cb_r));
+    upshift_btn.fall(callback(upshift_test_cb_f));
+
+    downshift_btn.rise(callback(downshift_test_cb_r));
+    downshift_btn.fall(callback(downshift_test_cb_f));
+
+    drs_btn.rise(callback(drs_test_cb_r));
+    drs_btn.fall(callback(drs_test_cb_f));
+
+    settings_btn.rise(callback(settings_test_cb_r));
+    settings_btn.fall(callback(settings_test_cb_f));
+
+    aux1_btn.rise(callback(aux1_test_cb_r));
+    aux1_btn.fall(callback(aux1_test_cb_f));
+
+    aux2_btn.rise(callback(aux2_test_cb_r));
+    aux2_btn.fall(callback(aux2_test_cb_f));
+}
+
+// Goes into button test mode
+void test_mode_enable_handler() {
+    btn_test_counter ++;
+    // Press settings 5 times for btn test mode
+    if(btn_test_counter >= 5){
+        queue.call(test_mode);
+    }
+}
+
+
+void send_can_message(bool drs, bool upshift, bool downshift, bool launch_ctl) {
+    unsigned char buffer[4] = {0, 0, 0, 0};
+    if(drs){
+        buffer[0] = 1;
+    }
+
+    if(upshift){
+        buffer[1] = 1;
+    }
+
+    if(downshift){
+        buffer[2] = 1;
+    }
+
+    if(launch_ctl){
+        buffer[3] = 1;
+    }
+    CANMessage msg(STEERING_WHEEL_ID, buffer, 4);
+
+    if(can.write(msg)){
+        return;
+    } else{
+        printf("Failed to send CAN message :(\n");
+    }
+}
+
+
+// Send upshift/downshift messages over CAN
+void upshift() {
+    {
+        ScopedLock <Mutex> lock(state.mutex); // Make sure the state isn't being otherwise modified or read while updating
+
+        send_can_message(state.drsEngaged, true, false, state.launchCtlEngaged);
```

Comment:
Would it make more sense to just pass shift type in here and handle this in the send function? Or add upshift and downshift to the state? Not super crazy about just straight up passing in t/f